crates/store/src/state.rs

```diff
@@ -783,61 +690,110 @@ impl State {
+        let highest_block_number = blocks.last().copied().unwrap_or(latest_block_number);
+        if highest_block_number > latest_block_number {
+            return Err(GetBlockInputsError::BatchBlockReferenceNewerThanLatestBlock {
+                highest_block_number,
+                latest_block_number,
             });
         }
 
-        // using current block number gets us the peaks of the chain MMR as of one block ago;
-        // this is done so that latest.chain_root matches the returned peaks
-        let chain_peaks =
-            inner.blockchain.peaks_at(latest.block_num().as_usize()).map_err(|error| {
-                GetBlockInputsError::FailedToGetMmrPeaksForForest {
-                    forest: latest.block_num().as_usize(),
-                    error,
-                }
-            })?;
-        let account_states = account_ids
+        // The latest block is not yet in the chain MMR, so we can't (and don't need to) prove its
+        // inclusion in the chain.
+        blocks.remove(&latest_block_number);
+
+        // Fetch the partial MMR at the state of the latest block with authentication paths for the
+        // provided set of blocks.
+        let partial_mmr = inner.blockchain.partial_mmr_from_blocks(&blocks, latest_block_number);
+
+        // Fetch witnesses for all acounts.
+        let account_witnesses = account_ids
             .iter()
             .copied()
             .map(|account_id| {
-                let ValuePath { value: account_hash, path: proof } =
-                    inner.account_tree.open(&LeafIndex::new_max_depth(account_id.prefix().into()));
-                Ok(AccountInputRecord { account_id, account_hash, proof })
+                let ValuePath {
+                    value: latest_state_commitment,
+                    path: proof,
+                } = inner.account_tree.open(&account_id.into());
+                (account_id, AccountWitness::new(latest_state_commitment, proof))
             })
-            .collect::<Result<_, AccountError>>()?;
+            .collect::<BTreeMap<AccountId, AccountWitness>>();
 
-        let nullifiers: Vec<NullifierWitness> = nullifiers
+        // Fetch witnesses for all nullifiers. We don't check whether the nullifiers are spent or
+        // not as this is done as part of proposing the block.
+        let nullifier_witnesses: BTreeMap<Nullifier, NullifierWitness> = nullifiers
             .iter()
+            .copied()
             .map(|nullifier| {
-                let proof = inner.nullifier_tree.open(nullifier);
-
-                NullifierWitness { nullifier: *nullifier, proof }
+                let proof = inner.nullifier_tree.open(&nullifier);
+                (nullifier, NullifierWitness::new(proof))
             })
             .collect();
 
-        let found_unauthenticated_notes =
-            self.get_note_authentication_info(unauthenticated_notes).await?;
+        // Release the lock.
+        std::mem::drop(inner);
```

Comment:
Would it make sense to have the above in a separate method? iiuc we avoid actual db access while we are locked, so we may as well?

As in, including from `.lock` until here.